api/src/main/java/org/openmrs/module/ModuleFactory.java

```diff
@@ -694,9 +709,32 @@ public static Module startModuleInternal(Module module, boolean isOpenmrsStartup
 					Context.removeProxyPrivilege("");
 				}
 				
-				// run module's optional liquibase.xml immediately after sqldiff.xml
-				log.debug("Run module liquibase: {}", module.getModuleId());
-				runLiquibase(module);
+				// Run version-change hooks + liquibase if necessary
+				String prevCoreVersion = getStoredCoreVersion() != null ? getStoredCoreVersion() : OpenmrsConstants.OPENMRS_VERSION_SHORT;
+				String currentCoreVersion = OpenmrsConstants.OPENMRS_VERSION_SHORT;
+
+				boolean forceSetup = Boolean.parseBoolean(
+					Context.getAdministrationService().getGlobalProperty("force.setup", "false"));
+
+				String prevModuleVersion = getStoredModuleVersion(module);
+				String currentModuleVersion = module.getVersion();
+
+				boolean versionChanged = forceSetup
+					|| versionChanged(prevCoreVersion, currentCoreVersion, prevModuleVersion, currentModuleVersion);
+				
+				if (versionChanged) {
+					runInSingleTransaction(() -> {
+						module.getModuleActivator().setupOnVersionChangeBeforeSchemaChanges(prevCoreVersion, prevModuleVersion);
```

Comment:
The idea of this method is to call it before running *ANY* liquibase changesets, including ones run by core so it should not be here rather in Context.checkForDatabaseUpdates...